src/seq.rs

```diff
@@ -403,114 +406,329 @@ pub fn sample_slice<R, T>(rng: &mut R, slice: &[T], amount: usize) -> Vec<T>
+/// `O(amount)` time.
 #[cfg(feature = "alloc")]
-fn sample_indices_inplace<R>(rng: &mut R, length: usize, amount: usize) -> Vec<usize>
+fn sample_indices_inplace<R>(rng: &mut R, length: u32, amount: u32) -> Indices
     where R: Rng + ?Sized,
 {
     debug_assert!(amount <= length);
-    let mut indices: Vec<usize> = Vec::with_capacity(length);
+    let mut indices: Vec<u32> = Vec::with_capacity(length as usize);
     indices.extend(0..length);
     for i in 0..amount {
-        let j: usize = rng.gen_range(i, length);
-        indices.swap(i, j);
+        let j: u32 = rng.gen_range(i, length);
+        indices.swap(i as usize, j as usize);
     }
-    indices.truncate(amount);
-    debug_assert_eq!(indices.len(), amount);
-    indices
+    indices.truncate(amount as usize);
+    debug_assert_eq!(indices.len(), amount as usize);
+    Indices::from(indices)
 }
 
-
-/// This method performs a partial fisher-yates on a range of indices using a
-/// `HashMap` as a cache to record potential collisions.
+/// Randomly sample exactly `amount` indices from `0..length`, using a
+/// dynamically-cached partial Fisher-Yates method.
 ///
-/// The cache avoids allocating the entire `length` of values. This is especially useful when
-/// `amount <<< length`, i.e. select 3 non-repeating from `1_000_000`
+/// The cache avoids allocating the entire `length` of values. This is
+/// especially useful when `amount <<< length`; e.g. selecting 3 non-repeating
+/// values from `1_000_000`. The algorithm is `O(amount)` time and memory,
+/// but due to overheads will often be slower than other approaches.
 #[cfg(feature = "alloc")]
-fn sample_indices_cache<R>(
-    rng: &mut R,
-    length: usize,
-    amount: usize,
-) -> Vec<usize>
+fn sample_indices_cache<R>(rng: &mut R, length: usize, amount: usize) -> Indices
     where R: Rng + ?Sized,
 {
     debug_assert!(amount <= length);
     #[cfg(feature="std")] let mut cache = HashMap::with_capacity(amount);
     #[cfg(not(feature="std"))] let mut cache = BTreeMap::new();
-    let mut out = Vec::with_capacity(amount);
+    let mut indices = Vec::with_capacity(amount);
     for i in 0..amount {
         let j: usize = rng.gen_range(i, length);
 
-        // equiv: let tmp = slice[i];
-        let tmp = match cache.get(&i) {
-            Some(e) => *e,
+        // get the current values at i and j ...
+        let x_i = match cache.get(&i) {
+            Some(x) => *x,
             None => i,
         };
-
-        // equiv: slice[i] = slice[j];
-        let x = match cache.get(&j) {
+        let x_j = match cache.get(&j) {
             Some(x) => *x,
             None => j,
         };
 
-        // equiv: slice[j] = tmp;
-        cache.insert(j, tmp);
-
-        // note that in the inplace version, slice[i] is automatically "returned" value
-        out.push(x);
+        // ... and swap them
+        cache.insert(j, x_i);
+        indices.push(x_j);  // push at position i
```

Comment:
For `amount <<< length` the following should be faster I would think:

```rust
#[cfg(feature="std")] let mut cache = HashSet::with_capacity(amount);
#[cfg(not(feature="std"))] let mut cache = BTreeSet::new();
let distr = Uniform::new(0, length);
for _ in 0..amount {
    let mut pos = distr.sample(rng);
    while !cache.insert(pos) {
        pos = distr.sample(rng);
    }
    indices.push(pos);
}
```

Reply:
Yes, this seems to cut run-time by around half (even for huge values like `misc_sample_indices_1M_of_10G` and ranges close to the cut-off like `misc_sample_indices_140k_of_10M`).

I'll have to re-do my selection heuristics...